docs/design/openstack/networking-infrastructure.md

```diff
@@ -0,0 +1,71 @@
+
+The list of `etcd` hosts included in the SRV record is based on the list of
+control plane nodes currently running.
+
+The IP addresses that the `etcd-NNN` host records resolve to comes from the
+mDNS advertisement sent out by the `mdns-publisher` on that control plane node.
+
+## Virtual IP's
+
+We use virtual IP addresses, VIPs, managed by keepalived
+to provide high availability access to essential APIs and services. For more info
+on how this works, please read about what [keepalived is](https://www.keepalived.org/) and
+about the underlying [VRRP algorithm](https://en.wikipedia.org/wiki/Virtual_Router_Redundancy_Protocol)
+that it runs. In our current implementation, we have 3 highly available VIPs that we manage: 
+Ingress, which handles requests to services managed by OpenShift, DNS, which handles internal dns requests, and API, which handles requests to the openshift API. Our VIP addresses are chosen and validated from the nodes subnet in the openshift
+installer, however our VIPs along with our internal dns and loadbalancing is managed
+by the [machine-config-operator(MCO)](https://github.com/openshift/machine-config-operator/tree/master/docs).
+The MCO has been configured to run static pods on the Bootstrap, Master, and Worker Nodes that
+run our internal networking infrastructure. Files run on the bootstrap node can be found 
+[here](https://github.com/openshift/machine-config-operator/tree/master/manifests/openstack).
+Files run on both master and worker nodes can be found 
+[here](https://github.com/openshift/machine-config-operator/tree/master/templates/common/openstack/files).
+Files run on only master nodes can be found 
+[here](https://github.com/openshift/machine-config-operator/tree/master/templates/master/00-master/openstack/files).
+Lastly, files run only on worker nodes can be found 
+[here](https://github.com/openshift/machine-config-operator/tree/master/templates/worker/00-worker/openstack/files).
+
+## Infrastructure Walkthrough
+
+The bootstrap node is responsible for running temporary networking infrastructure while the Master
+nodes are still coming up. The bootstrap node will run a coredns instance, as well as 
+keepalived. While the bootstrap node is up, it will have priority running the API and DNS
+VIPs.
+
+The Master nodes run dhcp, haproxy, coredns, mdns publisher, and keepalived. Haproxy loadbalances incoming requests 
+to the api across all running masters. It also runs a stats and healthcheck server. Keepalived manages all 3 VIPs on the master, where each
+master has an equal chance of being assigned one of the VIPs. Be aware, the bootstrap node has the highest priority for hosting the DNS
+and API VIPs, so they will point to addresses there for as long as they can. Once the master nodes have stood up a fully functioning control plane, 
+the bootstrap node will be torn down by the installer. Keepalived implements periodic health checks to make sure
```

Comment:
This section is (IIRC) not entirely correct. Initially, bootstrap does have a higher priority than the master's. However, once a master passes the health check, its priority increases and the bootstrap -> master VIP switch happens then.

We do *not* rely on the bootstrap teardown for the API VIP transition.

This is the initial master priority: https://github.com/openshift/machine-config-operator/blob/db0210f8bb30117bcabd9fa74d8e557d09f787a2/templates/master/00-master/openstack/files/openstack-keepalived-keepalived.yaml#L27

This is the weight based on the health check: https://github.com/openshift/machine-config-operator/blob/db0210f8bb30117bcabd9fa74d8e557d09f787a2/templates/master/00-master/openstack/files/openstack-keepalived-keepalived.yaml#L9

The weight gets added to the priority, turning it from 40 to 90.

Bootstrap's priority is 50 and there are no other checks that would increase it: https://github.com/openshift/machine-config-operator/blob/db0210f8bb30117bcabd9fa74d8e557d09f787a2/manifests/openstack/keepalived.conf.tmpl#L11

See: https://www.keepalived.org/manpage.html

> A positive weight means that an OK status will  add  <weight>  to
> the priority of all VRRP instances which monitor it. On the opposite, a
> negative weight will be subtracted from the initial priority in case of
> insufficient processes.

/cc @celebdor or @bcrochet to make sure I'm not making this up.